hazelcast/src/main/java/com/hazelcast/sql/SqlResult.java

```diff
@@ -82,6 +84,23 @@ default boolean isRowSet() {
     @Override
     Iterator<SqlRow> iterator();
 
+    /**
+     * Returns a stream of result rows.
+     * <p>It uses internally {@link #iterator()} method, so it cannot be called twice.</p>
+     *
+     * @throws IllegalStateException if the method is invoked more than once or
+     *    if this result doesn't have rows
+     * @throws HazelcastSqlException in case of an SQL-related error condition
+     *
+     * @return Stream of result rows
+     *
+     * @since 5.3
+     */
+    @Nonnull
+    default Stream<SqlRow> stream() {
+        return StreamSupport.stream(spliterator(), false);
+    }
```

Comment:
I don't understand why this default method doesn't exist in the `Iterable` interface.

Reply:
LMGTFY: https://stackoverflow.com/a/23177907

Reply:
argument with specialised iterator types does not apply to us. I wonder if adding this method would promote less efficient (but easier) usage. Also users will not be able to use this method in a fully fluent way because `SqlResult` has to be closed

Reply:
Java architects' defenses on these type of poor design choices are really funny. They are reluctant to get rid of primitive/boxed duality (by burying auto-(un)boxing deeper) and they use this as a counter-argument in order not to implement some convenience methods. By the way, I upvoted [this][1] and [this][2].

To get rid of try-with-resources, we can return `StreamSupport.stream(spliterator(), false).onClose(this::close)`, but try-with-resources is the only reliable way to release resources in case a `RuntimeException` is thrown. Closing `SqlResult` in the end of `iterator()`, i.e. when `hasNext()` returns false, is also a bad idea since the iterator doesn't have to be used to the end.

[1]: https://stackoverflow.com/questions/23114015/why-does-iterablet-not-provide-stream-and-parallelstream-methods/23177907#comment44698509_23177907
[2]: https://stackoverflow.com/questions/23114015/why-does-iterablet-not-provide-stream-and-parallelstream-methods/23177907#comment63452496_23177907

Reply:
to say "Java is not perfect" would be a huge understatement, but we have to live with it

Reply:
Yeah I know it's not perfect - still needs try-with-resources - but it's simpler for queries where you expect not many result rows (e.g. with `limit`), similar to Mongo's `.into()` function